### PR TITLE
check subcommand should force rebuild

### PR DESCRIPTION
**Pull request by JoshMcguigan**
This builds on the work from #5944, and pushed forward rust-lang/rust-clippy#2604. If ensures cargo check always does a rebuild.

The one thing I'm not sure of is whether this should be optional behind a command line flag? My initial thoughts are that most users would expect by default that `cargo check` always worked in this way (rebuilding and showing warnings every time it is ran), but that users running `cargo check` on a file system watcher might have some issues.

### CODE REVIEW

**Comment by JoshMcguigan** [comment]
It looks like three tests failed, but only on nightly. I'm going to look into this, but any ideas of what could be the cause here? 

```
profile_targets::profile_selection_check_all_targets
profile_targets::profile_selection_check_all_targets_release
profile_targets::profile_selection_check_all_targets_test
```

**Comment by JoshMcguigan** [comment]
The latest commit should fix these tests, assuming we want to remove the early return for `!nightly()`, and also assuming we want the default behavior of `cargo check` to force a rebuild. Some of this would need to be reverted if the new `cargo check` behavior goes behind a command line option.

**Comment by dwijnand** [comment]
> most users would expect by default that `cargo check` always worked in this way (rebuilding and showing warnings every time it is ran)

I don't think most users expect `cargo check` to rebuild, quite the opposite I think they expect it to be fast and avoid rebuilding.  They might _want_ to see warnings every time (I'm with them) but, again, I don't think they expect it, given how long `cargo check` has existed by now.

If we agree `cargo check` should redisplay warnings then we should find a way to cache and redisplay the warnings every time.  If not this should be behind a flag. (`--rebuild`?)

**Comment by BatmanAoD** [comment]
@dwijnand I was pretty surprised to discover that `cargo check` doesn't re-display errors. But since it doesn't "build" in the first place, I wouldn't expect it to "rebuild" either way!

As long as this PR doesn't change the behavior of `cargo build` (i.e. in `cargo build; cargo check; cargo build`, the second build would still be a no-op), then I don't understand why this change would be surprising to users.

**Comment by dwijnand** [comment]
> As long as this PR doesn't change the behavior of `cargo build` (i.e. in `cargo build; cargo check; cargo build`, the second build would still be a no-op), then I don't understand why this change would be surprising to users.

It's surprising to users not because of any interactions with `cargo build` but because now every time they ran `cargo check` their entire crate would recompiled, making `cargo check` _worst_ than `cargo build`, given a big enough crate.

**Comment by JoshMcguigan** [comment]
> If we agree `cargo check` should redisplay warnings then we should find a way to cache and redisplay the warnings every time.

Yes, I should have said users expect to see the error messages each time, so I think cacheing these warnings would be great. Any tips on implementing this cacheing? I'm new to the cargo codebase so I'm not sure how cacheing is typically handled.

**Comment by dwijnand** [comment]
I don't know exactly but cargo process calls rustc, and does so via a rustc wrapper, so finding that will get you half way to before it executes it.

Then you need to go to the place where it uses the fact that the build artefacts are "fresh" to avoid compile to display the warnings you cached.

If I can find more specific code locations I'll recomment.

**Comment by sanmai-NL** [comment]
@dwijnand: `cargo check` shouldn't compile, it should one perform the first few steps in the build pipeline, as documented at https://rust-lang-nursery.github.io/edition-guide/rust-2018/cargo-and-crates-io/cargo-check-for-faster-checking.html. Maybe those first steps take a significant amount of time, but insofar that's unacceptable that'd be an issue in itself.

**Comment by sanmai-NL** [comment]
@BatmanAoD:
```sh
cargo build; cargo check; cargo build
```
should behave differently, `cargo check` shouldn't reuse `cargo build` results at all. That's why it's called 'check': the main use case for the check-subcommand is to check e.g. syntax again, irrespective of whether some build succeeded earlier and whether the source code/build config has changed in the meantime.

**Comment by dwijnand** [comment]
> @dwijnand: `cargo check` shouldn't compile, it should one perform the first few steps in the build pipeline, as documented at [rust-lang-nursery.github.io/edition-guide/rust-2018/cargo-and-crates-io/cargo-check-for-faster-checking.html](https://rust-lang-nursery.github.io/edition-guide/rust-2018/cargo-and-crates-io/cargo-check-for-faster-checking.html). Maybe those first steps take a significant amount of time, but insofar that's unacceptable that'd be an issue in itself.

Right, I said "compile" as short-hand, but you're right it stops earlier in the pipeline, running (what's commonly referred to as) the frontend of the compiler.

> @BatmanAoD:
> 
> ```shell
> cargo build; cargo check; cargo build
> ```
> 
> should behave differently, `cargo check` shouldn't reuse `cargo build` results at all. That's why it's called 'check': the main use case for the check-subcommand is to check e.g. syntax again, irrespective of whether some build succeeded earlier and whether the source code/build config has changed in the meantime.

`cargo check` actually doesn't reuse `cargo build`'s results, which is tracked in https://github.com/rust-lang/cargo/issues/3501.  I disagree that 'check' shouldn't make use of previous results, but I agree it should redisplay warnings, either by default or behind a flag. (`--show-warnings`? `--show-lints`?)

**Comment by sanmai-NL** [comment]
@dwijnand: I agree it should redisplay warnings then, but by default, since I claim is would be the expected  behavior. Since running the first steps in the compilation pipeline is supposed to be low-cost, the desired default behavior would be easily implemented by not reusing previous build results.

**Comment by dwijnand** [comment]
> Since running the first steps in the compilation pipeline is supposed to be low-cost, the desired default behavior would be easily implemented by not reusing previous build results.

I think it's low-cost compared to a full `build`, but it would be slow compared to the status quo in which cargo uses freshness checks.

**Comment by sanmai-NL** [comment]
But it's called 'check', when you run that explicitly, it's reasonable some cost is incurred. To the contrary, it's not reasonable e.g. `cargo check` doesn't redisplay its messages because of caching.

**Comment by dwijnand** [comment]
> But it's called 'check', when you run that explicitly, it's reasonable some cost is incurred. To the contrary, it's not reasonable e.g. `cargo check` doesn't redisplay its messages because of caching.

While I can agree with the later, I'd consider it a performance regression if `cargo check` in a big codebase where no file was touched to take too long.

I'm optimistic we can find a way to cache and uncache warnings without changing the freshness behaviour.

**Comment by BatmanAoD** [comment]
> @JoshMcguigan :

> Any tips on implementing this cacheing? I'm new to the cargo codebase so I'm not sure how cacheing is typically handled.

it looks like Eric Huss has some pointers; see the links in his comment here: https://internals.rust-lang.org/t/clippy-behavior-affected-by-incremental-compilation/9436/2?u=batmanaod

@sanmai-NL : Yes, I was asking specifically about the *last* part of the sequence, the second `cargo build`. My point was that `cargo check` should never invalidate the `cargo build` cache, *regardless* of whether or not the `build` cache has any effect on the `check` behavior. (I don't know if this PR actually *would* cause the `build` cache to be invalidated, but it's something I'd want to check to be sure.)

> @dwijnand :

> `cargo check` actually doesn't reuse `cargo build`'s results

...hold on, that's surprising. If `cargo check` doesn't reuse the build results, why doesn't in re-analyze files that have already been compiled?

**Comment by dwijnand** [comment]
It uses its own results not build's results.

**Comment by JoshMcguigan** [comment]
After further review of the way clippy integrates with cargo (https://github.com/rust-lang/rust-clippy/blob/master/src/main.rs#L103-L111), it looks like clippy actually does need a way to ensure rustc will actually be invoked. This means cacheing the results of a rustc invocation within cargo and re-displaying them does not solve the clippy use case (clippy team, please let me know if I am misunderstanding here). 

I propose this PR can add a `--force-rebuild` CLI flag to `cargo check` to solve the clippy use-case, and a later PR can implement cacheing/re-displaying warning messages when the user runs something like `cargo check; cargo check`.

**Comment by dwijnand** [comment]
LGTM.  The Cargo.toml isn't needed in the test (one is generated automatically).  Also, maybe we should add --force-rebuild to `cargo build`.  But I'm happy to land this as is.

I'd like to hear what @alexcrichton or @ehuss think, though, before landing this.

**Comment by JoshMcguigan** [comment]
> The Cargo.toml isn't needed in the test (one is generated automatically)

Thanks, I'll remove this.

> Also, maybe we should add --force-rebuild to `cargo build`

My primary concern is in the name of the arg. It only rebuilds the crate, but not its dependencies. It seems like that may not be intuitive if applied to `cargo build`, but I'm not sure what the use case would be for `cargo build --force-rebuild` so I can't say for sure. Along the same lines, perhaps `cargo check --force-recheck` makes more sense than `cargo check --force-rebuild`. 

Not to muddy the waters here I just want to make sure this is considered.

File: src/bin/cargo/commands/check.rs [hunk 1]
```diff
@@ -26,6 +26,7 @@ pub fn cli() -> App {
             "Check all targets",
         )
         .arg_release("Check artifacts in release mode, with optimizations")
+        .arg_force_rebuild("Force rebuild of this crate")
```
Comment (Nemo157) [inline src/bin/cargo/commands/check.rs]: What happens if you pass multiple crates to be checked in a workspace (`cargo check --all --force-rebuild`) or multiple targets in a lone crate (`cargo check --all-targets --force-rebuild`)?

Comment (JoshMcguigan) [reply src/bin/cargo/commands/check.rs]: Thanks for bringing this up, I wasn't sure of what the wording should be here. Perhaps this should say `Force rebuild of this workspace`?

Comment (Nemo157) [reply src/bin/cargo/commands/check.rs]: "the selected target(s)" or "the selected package(s)" maybe?

**Comment by Nemo157** [comment]
I just tested this in a complex workspace of mine and there seems to be some bugs in how the `force_rebuild` flag is actually handled by build config. Probably not a blocker for getting it exposed in the CLI, but I'll try and remember to open a bug ticket in a few days.

There are 11 crates in this workspace, 2 are excluded so there should be 9 being forced to rebuild, but the only ones that I have seen consistently included are `embrio-async-dehygiene` and `local`. These two are the only crates that aren't dependencies of other crates in the workspace, so it seems like the logic that excludes dependencies from being forced is also able to exclude packages that were explicitly selected:

```
> ../check-force-rebuild/target/debug/cargo check --all --exclude embrio-nrf51 --exclude pca10031 --force-rebuild
    Checking embrio-native v0.1.0 (/Users/nemo157/sources/embrio-rs/embrio-native)
    Checking embrio-async-dehygiene v0.1.0 (/Users/nemo157/sources/embrio-rs/embrio-async/dehygiene)
    Checking embrio v0.1.0 (/Users/nemo157/sources/embrio-rs/embrio)
    Checking hello v0.0.0 (/Users/nemo157/sources/embrio-rs/examples/apps/hello)
    Checking local v0.0.0 (/Users/nemo157/sources/embrio-rs/examples/local)
    Finished dev [unoptimized + debuginfo] target(s) in 1.17s

> ../check-force-rebuild/target/debug/cargo check --all --exclude embrio-nrf51 --exclude pca10031 --force-rebuild
    Checking embrio-native v0.1.0 (/Users/nemo157/sources/embrio-rs/embrio-native)
    Checking embrio-async-dehygiene v0.1.0 (/Users/nemo157/sources/embrio-rs/embrio-async/dehygiene)
    Checking local v0.0.0 (/Users/nemo157/sources/embrio-rs/examples/local)
    Finished dev [unoptimized + debuginfo] target(s) in 0.99s
```

File: src/bin/cargo/commands/check.rs (hunk 1 above)
Comment (dwijnand) [reply src/bin/cargo/commands/check.rs]: Selected targets probably works, implying package selection too.

**Comment by JoshMcguigan** [comment]
@Nemo157 I was able to recreate your bug with the test below. If this is an appropriate recreation of your scenario, it looks like that bug is unrelated to `--force-rebuild`.

I'm happy to try tracking this down, but do we want to hold this PR on fixing this, or should this be handled as a separate issue/PR?

```rust
// Confirm that --all and --exclude flags are correctly combined
#[test]
fn check_all_exclude() {
    let p = project()
        .file(
            "Cargo.toml",
            r#"
            [package]
            name = "foo"
            version = "0.0.1"
            authors = []

            [workspace]
            [dependencies]
            b = { path = "b" }
        "#,
        )
        .file("src/main.rs", "fn main() {}")
        .file("examples/a.rs", "fn main() {}")
        .file("tests/a.rs", "")
        .file("src/lib.rs", "")
        .file("b/Cargo.toml", &basic_manifest("b", "0.0.1"))
        .file("b/src/main.rs", "fn main() {}")
        .file("b/src/lib.rs", "")
        .file("c/Cargo.toml", &basic_manifest("c", "0.0.1"))
        .file("c/src/main.rs", "fn main() {}")
        .file("c/src/lib.rs", "")
        .file("d/Cargo.toml", &basic_manifest("d", "0.0.1"))
        .file("d/src/main.rs", "fn main() {}")
        .file("d/src/lib.rs", "")
        .build();

    p.cargo("check --all --exclude d")
        .with_stderr_does_not_contain("[CHECKING] d [..]")
        .with_stderr_contains("[CHECKING] foo [..]")
        .with_stderr_contains("[CHECKING] b [..]")
        .with_stderr_contains("[CHECKING] c [..]")
        .run();
}
```

**Comment by Nemo157** [comment]
I took a glance through the code and it looks to me that the issue is that if you hit https://github.com/rust-lang/cargo/blob/55895e7125c3eecd077d7f0ce811e093c2b660d8/src/cargo/core/compiler/mod.rs#L180 with a dependency that is also one of the top level targets then because of https://github.com/rust-lang/cargo/blob/55895e7125c3eecd077d7f0ce811e093c2b660d8/src/cargo/core/compiler/mod.rs#L137-L139 it will not notice that that target is meant to be rebuilt (I definitely did not do enough investigation to confirm this, but if the order that the top-level targets have their compile jobs generated in is non-deterministic then this explains the behaviour I was seeing very well). If I'm correct then this looks like an issue that's been present since `force_rebuild` was added for `cargo fix`, so IMO would be best fixed as a separate change.

---

I think your test case is behaving correctly, from the [docs on `workspace.members`]():

```toml
# Optional key, inferred from path dependencies if not present.
# Additional non-path dependencies that should be included must be given here.
```

So only `b` is part of the workspace and it correctly ignores checking `c` (and it appears `--exclude` with a package that is not in the workspace doesn't give an error/warning (#6678) for `d`).

**Comment by JoshMcguigan** [comment]
@Nemo157 Thanks for the detailed follow up. 

The test case below fails, and I believe covers your scenario. It seems like the `--exclude` was not a factor in the bug. 

Feel free to tag me when you make the bug report, as I'd like to attempt solving this one. Or if you'd like I can write up the bug report. That is assuming @dwijnand doesn't think this should be fixed before merging this PR. 

```rust
// Confirm that force rebuild works correctly with dependencies which are also a target
#[test]
fn check_force_rebuild_with_workspace_dependencies() {
    let p = project()
        .file(
            "Cargo.toml",
            r#"
            [workspace]
            members = ["a", "b", "c", "d"]
        "#,
        )
        .file(
            "a/Cargo.toml",
            r#"
            [package]
            name = "a"
            version = "0.1.0"

            [dependencies]
            b = { path = "../b" }
            c = { path = "../c" }
            d = { path = "../d" }
            e = { path = "../e" }
            f = { path = "../f" }
        "#,
        )
        .file("a/src/lib.rs", "")
        .file("b/Cargo.toml", &basic_manifest("b", "0.0.1"))
        .file("b/src/lib.rs", "")
        .file("c/Cargo.toml", &basic_manifest("c", "0.0.1"))
        .file("c/src/lib.rs", "")
        .file("d/Cargo.toml", &basic_manifest("d", "0.0.1"))
        .file("d/src/lib.rs", "")
        .file("e/Cargo.toml", &basic_manifest("e", "0.0.1"))
        .file("e/src/lib.rs", "")
        .file("f/Cargo.toml", &basic_manifest("f", "0.0.1"))
        .file("f/src/lib.rs", "")
        .build();

    p.cargo("check --all")
        .with_stderr_contains("[CHECKING] a [..]")
        .with_stderr_contains("[CHECKING] b [..]")
        .with_stderr_contains("[CHECKING] c [..]")
        .with_stderr_contains("[CHECKING] d [..]")
        .with_stderr_contains("[CHECKING] e [..]")
        .with_stderr_contains("[CHECKING] f [..]")
        .run();

    p.cargo("check --all --force-rebuild")
        .with_stderr_contains("[CHECKING] a [..]")
        .with_stderr_contains("[CHECKING] b [..]")
        .with_stderr_contains("[CHECKING] c [..]")
        .with_stderr_contains("[CHECKING] d [..]")
        .with_stderr_contains("[CHECKING] e [..]")
        .with_stderr_contains("[CHECKING] f [..]")
        .run();
}
```

**Comment by alexcrichton** [comment]
Thanks for the PR! If this ends up adding a new flag I think the only thing I'd add is that we'll want to be sure to put it behind a feature flag (like `-Z unstable-options`) since it's a new feature.

**Comment by Manishearth** [comment]
Clippy needs the ability to use it internally, would this affect stable
clippy?

**Comment by dwijnand** [comment]
Yeah, it wouldn't be useable by stable Clippy until it's a stabilised feature in Cargo.

**Comment by Manishearth** [comment]
Can we have some sort of environment variable hack to make this possible
again?

**Comment by dwijnand** [comment]
Again?  When was this possible before?

**Comment by Manishearth** [comment]
It wasn't, it was hypothetically possible if this PR landed as is :)

Doesn't have to happen in this PR, and an alternative for us is to just
help it stabilize faster.

**Comment by dwijnand** [comment]
The other alternative is to be like `cargo fix` and uplift `cargo clippy` into the cargo repo, where it can programmatically set `force_rebuild` without having to wait for `--force-rebuild` to be stable.

**Comment by Manishearth** [comment]
Oh we've been considering doing that anyway, actually, especially because
the interactions between cargo fix and cargo clippy are tricky.

I might get to work on that.

**Comment by JoshMcguigan** [comment]
> Thanks for the PR! If this ends up adding a new flag I think the only thing I'd add is that we'll want to be sure to put it behind a feature flag (like `-Z unstable-options`) since it's a new feature.

I just pushed a commit which takes care of this. I used the unstable `build-plan` flag as a guide, but I noticed that running `cargo run -- build --help` shows the `build-plan` option even without passing the `Z` to show unstable options. This behavior carries over in my implementation as well, not sure whether that is the intended behavior or not. However, if you actually try to use the `force-rebuild` flag without `-Z unstable-options` you'll get an error (this was confirmed with a unit test).

**Comment by JoshMcguigan** [comment]
The test failure is in the `meta_test_deep_trees_from_strategy` test, so I think it will pass if retried.

**Comment by JoshMcguigan** [comment]
@Nemo157 For some reason I am not able to duplicate your issue under `cargo fix` on the master branch. Somehow it only seems to affect `cargo check` as demonstrated in the failing test above (or I misconfigured the test for `cargo fix`). 

Given that, perhaps we want to reconsider whether the fix merges with this PR? For now I've pushed the fix in a commit to a different branch (https://github.com/rust-lang/cargo/commit/9b3138c3458416380bb41b60555cffeb758a6c41), but if everyone agrees I can merge that commit into this branch/PR.

**Comment by Nemo157** [comment]
`cargo fix` always working seems to be a side-effect of it implying `--all-targets`, maybe because it will include the `cfg(test)` mode build of each package which is not a dependency of the other targets. Using `check --force-rebuild --all --all-targets` in my local workspace correctly builds all local crates.

**Comment by alexcrichton** [comment]
I think it's fine for clippy to use "unstable" features of Cargo, although perhaps not literally unstable features as they are today. In any case clippy, as an official tool now, should be fine to have support in Cargo which isn't intended for stable consumption elsewhere. (note that this interface, should it want to exist, would probably need to be designed rather than tacked on to something else existing).

I also think that moving `cargo clippy` into Cargo itself is fine so long as `cargo clippy` doesn't get enough development to put a strain on this repository.

**Comment by Manishearth** [comment]
Yeah, cargo clippy is a very thin shim so it wouldn't change much.

**Comment by JoshMcguigan** [comment]
Let me know if there needs to be any additional work here before merging. The test failure is in the `meta_test_deep_trees_from_strategy` test, so I think it will pass if retried.

**Comment by benesch** [comment]
CI seems happy with this—is there anything preventing it from being merged?

**Comment by ehuss** [comment]
I don't think adding a flag of this nature is ultimately the direction that we want to go.  Caching shouldn't take very long to add once it is ready upstream.  cargo-clippy is in the process of moving into this repo where we will have some more flexibility for the clippy side of things.

I would like to propose the following:  Merge this PR **without any command-line flags** and **immediately stable**.  The intent would be that a caching implementation would remove this change in the near future.  I'm not convinced that "running rustc again" is significantly painful.  On cargo's source, for me with a warm cache, takes only a couple seconds.  On servo, it takes about 7s.  Clippy is a bit slower, but still reasonable.

@rfcbot fcp merge

If this fcp does not pass, I propose closing this PR instead and waiting for aforementioned items.  Adding an unstable flag here will likely have the same timeframe as the caching implementation.

**Comment by withoutboats** [comment]
I'm not sure about merging this. I have the experience (and maybe many others do) of habitually running commands in a sort of impulsive & subconscious way as I do my work. `git status` for example. If `cargo check` is one of these for someone (because normally it finishes instantaneously and only doesn't when theyve actually done something they want to check), this would be really disruptive for their work flow as they are now taken out of the flow by this command.

**Comment by joshtriplett** [comment]
@rfcbot reviewed

Note that while I do want to see the *effect* of this, I'd like to see the caching as well (both in the normal and clippy-enabled cases). I regularly do clippy/edit/clippy cycles and would like those to go as quickly as possible.

**Comment by ehuss** [comment]
Sounds like a reasonable concern.
@rfcbot concern habitual-check-performance

**Comment by joshtriplett** [comment]
@ehuss @withoutboats I do think this is a great argument for caching. No good reason to rebuild just to display results. (And ideally, "fix one bug and recheck to get the rest" should go as quickly as possible and not take the same amount of time as the first check.)

**Comment by Nemo157** [comment]
It has been 5 months since the last update, has there been any movement on getting caching of warnings?

> I regularly do clippy/edit/clippy cycles and would like those to go as quickly as possible.

> (And ideally, "fix one bug and recheck to get the rest" should go as quickly as possible and not take the same amount of time as the first check.)

Are these use-cases actually significantly affected by these changes? If you are editing the `lib` then `cargo check` will have to recheck every target anyway, if you are editing a bin/test/other target and running `cargo check --all-targets` then you will unnecessarily be rechecking the lib and other unchanged targets, but that doesn't seem like such a massive downside to block making `cargo check` more usable from the command-line (IMO).

**Comment by mati865** [comment]
> It has been 5 months since the last update, has there been any movement on getting caching of warnings?

Clippy issue can be considered as fixed by https://github.com/rust-lang/cargo/pull/6759. Regular `cargo clippy` is not fixed yet (it shows error only for the first time) but `cargo clippy-preview -Z unstable-options` which will internally replace `cargo clippy` at some point works well and shows warnings/error every time.

**Comment by Nemo157** [comment]
From my testing #6759 appears to essentially do what this change does and just recheck the targets every time without any caching of warnings (unless there's caching happening inside `clippy-driver` rather than in `cargo`)?

**Comment by BatmanAoD** [comment]
🎉

**Pull request closed without merging**